homeassistant/components/whirlpool/sensor.py

```diff
@@ -0,0 +1,180 @@
+"""The Washer/Dryer Sensor for Whirlpool Appliances."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+
+from whirlpool.washerdryer import MachineState, WasherDryer
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_UNKNOWN
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+
+from . import WhirlpoolData
+from .const import (
+    CYCLE_FUNC,
+    DOMAIN,
+    ICON_D,
+    ICON_W,
+    MACHINE_STATE,
+    TANK_FILL,
+    WASHER_STATE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def washer_state(washer: WasherDryer) -> str | None:
+    """Determine correct states for a washer."""
+
+    machine_state = washer.get_machine_state()
+    machine_cycle = None
+    for count, func in zip(range(len(CYCLE_FUNC)), CYCLE_FUNC):
+        if func(washer):
+            machine_cycle = WASHER_STATE.get(count)
+            break
+
+    if washer.get_attribute("Cavity_OpStatusDoorOpen") == "1":
+        machine_cycle = "Door open"
+        return machine_cycle
```

Comment:
Move this to the top of the function since it doesn't really depend on cycle functions? Also no need to assign the variable just to return it:

```suggestion
    if washer.get_attribute("Cavity_OpStatusDoorOpen") == "1":
        return machine_cycle
```